peer/peer.go

```diff
@@ -0,0 +1,1942 @@
+	if strings.HasSuffix(address, ".onion") {
+		return cfg.Oniondial(network, address)
+	}
+	return cfg.Dial(network, address)
+}
+
+// minUint32 is a helper function to return the minimum of two uint32s.
+// This avoids a math import and the need to cast to floats.
+func minUint32(a, b uint32) uint32 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// newNetAddress attempts to extract the IP address and port from the passed
+// net.Addr interface and create a bitcoin NetAddress structure using that
+// information.
+func newNetAddress(addr net.Addr, services wire.ServiceFlag) (*wire.NetAddress, error) {
+	// addr will be a net.TCPAddr when not using a proxy.
+	if tcpAddr, ok := addr.(*net.TCPAddr); ok {
+		ip := tcpAddr.IP
+		port := uint16(tcpAddr.Port)
+		na := wire.NewNetAddressIPPort(ip, port, services)
+		return na, nil
+	}
+
+	// addr will be a socks.ProxiedAddr when using a proxy.
+	if proxiedAddr, ok := addr.(*socks.ProxiedAddr); ok {
+		ip := net.ParseIP(proxiedAddr.Host)
+		if ip == nil {
+			ip = net.ParseIP("0.0.0.0")
+		}
+		port := uint16(proxiedAddr.Port)
+		na := wire.NewNetAddressIPPort(ip, port, services)
+		return na, nil
+	}
+
+	// For the most part, addr should be one of the two above cases, but
+	// to be safe, fall back to trying to parse the information from the
+	// address string as a last resort.
+	host, portStr, err := net.SplitHostPort(addr.String())
+	if err != nil {
+		return nil, err
+	}
+	ip := net.ParseIP(host)
+	port, err := strconv.ParseUint(portStr, 10, 16)
+	if err != nil {
+		return nil, err
+	}
+	na := wire.NewNetAddressIPPort(ip, uint16(port), services)
+	return na, nil
+}
+
+// outMsg is used to house a message to be sent along with a channel to signal
+// when the message has been sent (or won't be sent due to things such as
+// shutdown)
+type outMsg struct {
+	msg      wire.Message
+	doneChan chan struct{}
+}
+
+// Stats is the collection of stats related to a peer
+type Stats struct {
+	StatsMtx           sync.Mutex // protects all statistics below here.
+	versionKnown       bool
+	protocolVersion    uint32
+	versionSent        bool
+	verAckReceived     bool
+	services           wire.ServiceFlag
+	timeOffset         int64
+	timeConnected      time.Time
+	lastSend           time.Time
+	lastRecv           time.Time
+	bytesReceived      uint64
+	bytesSent          uint64
+	userAgent          string
+	startingHeight     int32
+	lastBlock          int32
+	lastAnnouncedBlock *wire.ShaHash
+	lastPingNonce      uint64    // Set to nonce if we have a pending ping.
+	lastPingTime       time.Time // Time we sent last ping.
+	lastPingMicros     int64     // Time for last ping to return.
+}
+
+// ShaFunc is a function which returns a block sha, height and error
+// It is used as a callback to get newest block details.
+type ShaFunc func() (sha *wire.ShaHash, height int64, err error)
+
+// Peer provides a bitcoin peer for handling bitcoin communications.  The
```

Comment:
This comment needs to be updated to reflect reality.  Some of it will need to move to the block manager.